src/vision/status_quo/barbara_wishes_for_easy_runtime_switch.md

```diff
@@ -0,0 +1,68 @@
+# 😱 Status quo stories: Barbara wishes for an easy runtime switch
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from
+real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async
+Rust programmers face today.
+
+## The story
+
+[Barbara] has been working on an async codebase for the past 5 years. It is extremely mature and
+quite large (in the millions of lines of code). The product has traditionally been run on the server
+only, but now they want to move large parts to the edge. After some investigation, they realize that
+the open-source async runtime they have been using (tokio), while great for server workloads, is
+not very performant on the low-end edge devices.
+
+Tokio makes tradeoffs in memory for performance and while performance is still a concern on
```

Comment:
Tokio has low-resource modes, it would be helpful to be more precise around what specifically Tokio does that doesn't work.

Reply:
It seems like this is a contrived example specifically intended to describe some arbitrary capability tokio doesn't support, along the lines of the second bullet in the morals of the story section. Maybe TTCP from [alan switches runtimes](https://github.com/rust-lang/wg-async-foundations/blob/master/src/vision/shiny_future/alan_switches_runtimes.md) would provide a clearer example?

Reply:
If I understand your point, you are suggesting that the reason for wanting to switch is hypothetical and not necessarily representative of an actual story? I also looked at the TTCP story. In that case, I am able to translate TTCP to a real use case (e.g. QUIC). In this case, I am not. The story describes starting with Tokio, but then hitting some hypothetical problem.

The reason I bring this up is because **if** there is a real blocker w/ Tokio, then the "shiny future" story could be to fix the problem and then there is no reason to switch :) but the way this status quo story is written doesn't leave that option open.

Reply:
Exactly. Although I didn't write this story, I have similar concerns with the status quo that start from the assumption outlined in the second bullet point of the moral of the story section: That one of the _benefits_ of async rust's runtime being external to std is that there can and will be different user needs which don't need to be met by one single executor. If the tokio design goal is to support all possible use cases of a [std] async executor as well as a purpose built executor could, that would a) be super exciting and b) would universally obviate the "switching runtimes" concern by treating any need for a different executor as an issue for tokio to introduce the right extension point

Reply:
I can say that Tokio would like to support as many use cases as it can :) I believe that something like [Embassy](https://github.com/embassy-rs/embassy) will always be out of scope, but I also believe a user of Embassy would have to build their app assuming Embassy from the start.

Reply:
> Tokio has low-resource modes, it would be helpful to be more precise around what specifically Tokio does that doesn't work.

We didn't know of that. :) Could you please point to the relevant docs? As @jbr pointed out (I think?), while the point of this story was to cover the problems of switching a runtime, we may not have been particularly successful at coming up with compelling reasons for Barbara to want to switch the runtime.

@nikomatsakis Perhaps we should switch back to our hypothetical runtime?

Reply:
The main piece of relevant documentation is the [`tokio::runtime::Builder::new_current_thread`][1] function, which lets you create a runtime that doesn't spawn any threads. Additionally, the `#[tokio::main]` macro also has a configuration option for using the current thread runtime, which you can find on the documentation for the macro.

[1]: https://docs.rs/tokio/1/tokio/runtime/struct.Builder.html#method.new_current_thread

Reply:
Since threads were addressed already elsewhere, I understood @carllerche meant something more than just threads with `Tokio has low-resource modes`. Hence my question.

Reply:
I don't think switching back to a hypothetical runtime really helps that much. The question is why people want to switch. I will say that I've talked to quite a few people who express a desire to *have the option* to switch and experiment with other runtimes. Perhaps the story should be tweaked to be Barbara hearing about a new runtime -- or wanting to implement her own to see if she could do better -- and being curious to see what it would entail, then realizing it would be tricky and not doing it. That seems to be more common in practice. =)

Reply:
@nikomatsakis hmm.. good point. 

> Perhaps the story should be tweaked to be Barbara hearing about a new runtime -- or wanting to implement her own to see if she could do better

Sure but neither of those sound like compelling enough reasons to put a lot of effort and time into this. Barbara will likely just give-up as soon as they realize it's not a trivial endeavor. Although the story doesn't have to be super realistic. :)

How about they needed to add a new feature, which would be a lot easier to add using some existing libraries that are closely tied to another runtime?

Reply:
What about Barbara's company making a native desktop version of her app and realizing that there are performance gains from using an os-specific runtime like grand central dispatch?

Reply:
Personally I think it's sufficient to just say that they were investigating what it would take to switch. Maybe she was asked to assess it for her job so she did a good job digging into all the details.

Reply:
As a bit of context - this also comes from teams I've talked to who have expressed the desire to not be locked into a specific runtime. We have many happy users of tokio and many happy users of custom runtimes, and there are many teams who are unsure where tokio will always be able to meet their needs or whether they will eventually want to switch off. 

In short, this story wasn't motivated by a real world situation that exists yet (after all, async Rust is fairly new and so we're just not at a point where this might appear). I think it's reasonable to want to ensure that switching off of a certain runtime is doable.

Reply:
I came up with an alternative motivation (something more likely IMO). Please have a look. If people still disagree, it can be changed again of course. :)